intel_adsp: move west sign from west flash to earlier west build

marc-hb opened the pull request:
Invoking `west sign` in `west build` accelerates twister because `west
build` is run in parallel, see rationale in superseded and very different (CMake-based) PR #52942
- https://github.com/zephyrproject-rtos/zephyr/pull/52942.

To maximize backwards compatibility:
- `west sign` is optional in `west build`
- `west flash` will sign (again) if any rimage --option is passed

This PR started with "v3", v1 and v2 happened in #54189. v3 is significantly smaller and simpler than v2

Should fix https://github.com/thesofproject/sof/issues/6917

Unlike https://github.com/zephyrproject-rtos/zephyr/pull/52942, this moves the rimage configuration logic not to CMake but to `west sign` configuration instead.

You should read and understand #52942 before looking at this.

Also take a look at https://github.com/thesofproject/sof/issues/6917

Signed-off-by: Marc Herbert <marc.herbert@intel.com>

cc:
- https://github.com/zephyrproject-rtos/west/issues/429

marc-hb committed:
west: sign: add new west config [rimage].extra-args and a default key

Make RIMAGE_SIGN_KEY a CMake CACHE variable so `west sign` can find it
and use it as a default value.

Fixes https://github.com/thesofproject/sof/issues/6917

Signed-off-by: Marc Herbert <marc.herbert@intel.com>

marc-hb committed:
west: sign: rimage: simplify rimage --tool-data code

Reduce duplication, no functional change except for a shorter log
statement with slightly less information.

This is required by the next commits.

Signed-off-by: Marc Herbert <marc.herbert@intel.com>

mbolivar-nordic reviewed (commented):
Thanks for the update. The overall design makes sense to me. I would ask that you think about a separate `rimage-extra-args` instead of `tool-extra-args`, so that we could add `imgtool-extra-args` in the future. This in turn would enable quick switching between `west sign --tool imgtool` and `west sign --tool rimage` in situations where users have to do both.

marc-hb committed:
west: sign: rimage: add default value '-c sof/rimage/config'

Don't force 99% of the users to pass the same value.

Signed-off-by: Marc Herbert <marc.herbert@intel.com>

marc-hb committed:
west: sign: add ability to read parameters from west config

No functional change yet, will be used in the next commits.

Signed-off-by: Marc Herbert <marc.herbert@intel.com>

marc-hb committed:
west: sign: add west config [sign].tool = rimage/imgtool

The --tool parameter is not required anymore thanks a [sign] entry in
west config like this one:

```
[sign]
tool = imgtool
```

Signed-off-by: Marc Herbert <marc.herbert@intel.com>

marc-hb committed:
west: sign: add west config [rimage].path = /path/to/rimage

Add a 3rd option besides --tool-path and $PATH

Signed-off-by: Marc Herbert <marc.herbert@intel.com>

marc-hb committed:
intel_adsp: move `west sign` from `west flash` to earlier `west build`

Invoking `west sign` in `west build` accelerates twister because `west
build` is run in parallel, see rationale in superseded and very
different (CMake-based) PR #52942.

To maximize backwards compatibility:
- `west sign` is optional in `west build`
- `west flash` will sign (again) if any rimage --option is passed

Signed-off-by: Marc Herbert <marc.herbert@intel.com>

marc-hb commented:
I'm guessing the failures are because CI does not have rimage. Maybe I'll make signing at build time optional, downgrade a missing rimage to a warning (and make sure there's a good error message at later flash time)

marc-hb commented:
I just discovered the pretty hard way that `CONFIG_CLEANUP_INTERMEDIATE_FILES=y` voids any twister performance benefit brought by this PR.

`CONFIG_CLEANUP_INTERMEDIATE_FILES=y` was added by commit 7dc1ef03b99 and a few, similar other commits.

Workaround:
```
# only the first time! Passing extra CMake arguments rebuilds from scratch always
west build -b intel_adsp_cavs15 samples/hello_world/ -- -DCONFIG_CLEANUP_INTERMEDIATE_FILES=n
west build 
```
`prj.conf` or some overlay can also do this.

The example above increases the build directory size from 15M to 16M.

marc-hb commented:
> I just discovered the pretty hard way[*] that CONFIG_CLEANUP_INTERMEDIATE_FILES=y voids any twister performance benefit brought by this PR.

- #55347 is also required to fix the incremental build but that one was easy and expected.

marc-hb committed:
west: sign: define filename constants earlier in rimage signer

Zero functional change, pure preparation for the next commit.

Signed-off-by: Marc Herbert <marc.herbert@intel.com>

marc-hb committed:
west: sign: clean any stale rimage output

We never want to leave stale outputs behind after failing.

Signed-off-by: Marc Herbert <marc.herbert@intel.com>

marc-hb committed:
west: sign: add new rimage option --if-tool-available

Moving `west sign` from `west flash` to `west build` for rimage has
multiple advantages (including a bit more consistency with
imgtool). However it makes `west build` fail when rimage is missing.

To avoid forcing every CI and developer who never used it before to
install rimage, make signing optional when passing new `west sign`
option --if-tool-available. A clear warning is printed.

Signed-off-by: Marc Herbert <marc.herbert@intel.com>

teburd commented:
This works for me

mbolivar-nordic commented on scripts/west_commands/sign.py:
```diff
@@ -9,12 +9,14 @@
 from west import log
 from west import manifest
 from west.util import quote_sh_list
+from west.configuration import config as west_config
```
This is deprecated; please use `WestCommand.config` instead. https://docs.zephyrproject.org/latest/develop/west/west-apis.html#westcommand

mbolivar-nordic commented on scripts/west_commands/sign.py:
```diff
@@ -423,11 +423,12 @@ class RimageSigner(Signer):
-                                '--tool-path {}: not an executable'.
-                                format(args.tool_path))
-            tool_path = args.tool_path
+        tool_path = args.tool_path if args.tool_path else config_get('rimage', 'path', None)
```
... i.e. replace this with `command.config.get('rimage.path')` instead of `config_get`

mbolivar-nordic commented on scripts/west_commands/sign.py:
```diff
@@ -506,9 +506,16 @@ def sign(self, command, build_dir, build_conf, formats):
-                     ['-o', out_bin] + conf_path_cmd + extra_ri_args +
-                     components)
+
+        sign_config_extra_args = config_get_words('rimage', 'extra-args', [])
```
... reimplement this on top of `command.config.get('rimage.extra-args')`

mbolivar-nordic commented on scripts/west_commands/sign.py:
```diff
@@ -506,9 +506,16 @@ def sign(self, command, build_dir, build_conf, formats):
+            cmake_default_key = cache.get('RIMAGE_SIGN_KEY')
+            extra_ri_args += [ '-k', str(sof_src_dir / 'keys' / cmake_default_key) ]
+
+        # Warning: the _args order matters! Think hard.
```
Surely a person who isn't familiar with this code would benefit from a better comment here :P

mbolivar-nordic commented:
Looks fine from `west sign` perspective; will leave the `rimage` portions for y'all to sort out

kv2019i commented on doc/develop/west/sign.rst:
```diff
@@ -94,3 +94,54 @@ file an issue for adding support.
+environment differences. Last and worst: new signing feature and options are
+impossible to use until more boilerplate code has been added in each layer.
+
+To avoid these issues, ``rimage`` parameters can bet set in ``west config``
+instead. Here's a ``workspace/.west/config`` example:
```
typoe: can bet set

kv2019i reviewed (dismissed):
+1 on the intel_adsp specific changes

nashif commented:
DNM: need to spend some time on this and test it proper.

aborisovich reviewed (dismissed):
Looks good. If all Zephyr CIs are ok with this change I'm all hands for it.

nashif commented on scripts/west_commands/runners/intel_adsp.py:
```diff
@@ -63,14 +47,6 @@ def capabilities(cls):
     def do_add_parser(cls, parser):
         parser.add_argument('--remote-host',
                             help='hostname of the remote targeting ADSP board')
-        parser.add_argument('--rimage-tool', default=DEFAULT_RIMAGE_TOOL,
```
removing all those options makes it less flexible and breaks existing users depending on those options.

marc-hb commented:
> removing all those [west flash] options makes it less flexible and breaks existing users depending on those options.

Yes, from the beginning the idea was to stop `west flash` from signing, transferred to `west build` instead. The much earlier #52942 was also dropping these signing options too with a... "little" difference: in #52942 `west flash` was _silently_ dropping signing options :-) 

In v4 bd4a5aecd63031 I think I found a simple way to preserve backwards compatibility and keep these options while achieving the other goals, especially signing in parallel => `west flash` now (re-)signs if any signing option was passed to it, otherwise it does not.

@nashif could you try this latest version? Thanks!

nashif commented:
you need to rebase.

marc-hb commented:
Thanks @nashif for spotting this, looks like it's not possible to be notified of new conflicts without... Slack! https://github.com/ploomber/contributing/issues/16

So v5 b3b327d6f50571a1e9142d4f7389408e988ad03a is a pure rebase, no code change besides resolving the minor `board.cmake` conflicts with the removal of older platforms in PR #56588. All tests passed in https://github.com/zephyrproject-rtos/zephyr/actions/runs/4652415270/jobs/8232619338?pr=54700 and internal tests passed as well.

The force-push lost all code reviews, would the previous reviewers be kind enough to re-approve? Sorry for the inconvenience.
